patterns/strategy.md

```diff
@@ -0,0 +1,175 @@
+}
+```
+
+## Advantages
+
+The main advantage is separation of concerns. For example, in this case `Report` does not know anything about specific
+implementations of `Json` and `Text`, whereas the output implementations does not care about how data is
+preprocessed, stored, and fetched.
+The only thing they have to know is context and a specific trait and method to implement,
+i.e,`Formatter` and `run`.
+
+## Disadvantages
+
+For each strategy there must be implemented at least one module, so number of modules
+increases with number of strategies.
+If there are many strategies to choose from then users have to know how strategies differ
+from one another.
+
+## Discussion
+
+In the previous example all strategies are implemented in a single file.
+Ways of providing different strategies includes:
+
+- All in one file (as shown in this example, similar to being separated as modules)
+- Separated as modules, E.g. `formatter::json` module, `formatter::text` module
+- Use compiler feature flags, E.g. `json` feature, `text` feature
+- Separated as crates, E.g. `json` crate, `text` crate
+
+Serde crate is a good example of the `Strategy` pattern in action. Serde allows [full customization](https://serde.rs/custom-serialization.html)
+of the serialization behavior by manually implementing `Serialize` and `Deserialize` traits for our type.
+For example, we could easily swap `serde_json` with `serde_cbor` since they expose similar methods.
+Having this makes the helper crate `serde_transcode` much more useful and ergonomic.
+
+However, we don't need to use traits in order to design this pattern in Rust.
+
+The following toy example demonstrates the idea of the Strategy pattern using Rust
+`closures`:
+
+```rust
+struct Adder;
+impl Adder {
+    pub fn add<F>(x: u8, y: u8, f: F) -> u8
+    where
+        F: Fn(u8, u8) -> u8,
+    {
+        f(x, y)
+    }
+}
+
+fn main() {
+    let arith_adder = |x, y| x + y;
+    let bool_adder = |x, y| {
+        if x == 1 || y == 1 {
+            1
+        } else {
+            0
+        }
+    };
+    let custom_adder = |x, y| 2 * x + y;
+
+    assert_eq!(9, Adder::add(4, 5, arith_adder));
+    assert_eq!(0, Adder::add(0, 0, bool_adder));
+    assert_eq!(5, Adder::add(1, 3, custom_adder));
+}
+
+```
+
+In fact, Rust already uses this idea for `Options`'s `map` method
+
+```rust
+fn main() {
+    let val = Some("Rust");
+
+    let len_strategy = |s: &str| s.len();
+    assert_eq!(4, val.map(len_strategy).unwrap());
+
+    let first_byte_strategy = |s: &str| s.bytes().next().unwrap();
+    assert_eq!(82, val.map(first_byte_strategy).unwrap());
+}
+```
+
+## See also
+
+- [Strategy Pattern](https://en.wikipedia.org/wiki/Strategy_pattern)
+- [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)<!-- markdown-link-check-disable -->
+- [Policy Based Design](https://en.wikipedia.org/wiki/Modern_C%2B%2B_Design#Policy-based_design)<!-- markdown-link-check-enable -->
```

Comment:
```suggestion
- [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection)
- [Policy Based Design](https://en.wikipedia.org/wiki/Modern_C++_Design#Policy-based_design)
```

Reply:
Can we try that and see if it works, otherwise we revert and keep the disabled link check

Reply:
(Would be an upstream issue if it works like this)